### DISCUSSION TITLE
thread array doesn't support fn foo() (int,int)

### DISCUSSION THREAD

**Discussion by timrichardson** [vlang/v#15300] [General]
22:39 $ v doctor
OS: linux, Ubuntu 22.04 LTS
Processor: 16 cpus, 64bit, little endian, AMD Ryzen 7 5800X 8-Core Processor
CC version: cc (Ubuntu 11.2.0-19ubuntu1) 11.2.0

getwd: /home/tim/learn_vlang
vmodules: /home/tim/.vmodules
vroot: /home/tim/v
vexe: /home/tim/v/v
vexe mtime: 2022-07-26 12:35:15
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.0 c976a69

Git version: git version 2.34.1
Git vroot status: weekly.2022.30-7-gc976a691
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 827f7452

**What did you do?**

I am surprised I can't do this:

```
module main

fn foo() (int, int) {
	return 1,2
}

fn main() {
	nbr_concurrent_tasks := 1000	
	mut tasks := []thread (int,int){}
	for i in 0 .. nbr_concurrent_tasks {
		tasks << go foo()
	}
	r := tasks.wait()
	println("$r")
}


```

**What did you expect to see?**
since foo() has a return type of (int,int) I expected that I could construct an array of tasks.

It seems that I can define a return type that is not really a type.
 
**What did you see instead?**
syntax error here: `mut tasks := []thread (int,int){}`

`issue_parallel.v:10:28: error: unexpected token `,`, expecting `)``

**Comment by spytheman** [comment]
As a workaround, you can return a named struct:
```v
struct Foo { f1 int f2 int }
fn foo() Foo { return Foo{1,2} }
fn main() {
	nbr_concurrent_tasks := 5
	mut tasks := []thread Foo{}
	for _ in 0 .. nbr_concurrent_tasks {
		tasks << go foo()
	}
	r := tasks.wait()
	println(r)
}
```

**Comment by timrichardson** [comment]
... or a two element array. 

However, is my report a bug that will be fixed?

**Comment by spytheman** [comment]
I would say that is more of a feature request, given that currently V does not have first class support for tuples.

**Comment by Apocryphon-X** [comment]
> currently V does not have first class support for tuples.

Maybe this is somehow related to #15083 (and the comment below the report)

**Comment by Wajinn** [comment]
Wondering if there is conceptual confusion between tuples and multiple returns.  They are not the same.  The tuple is a separate type of data structure, like so are structs, arrays, etc...  In languages like Vlang (and some others), the function is returning **separate values**, it is not any type of specific data structure.  Where with a language that has tuples they are returning a **single item** that contains multiple unnamed values.  

The tuple is also different from a struct, where even though you can have a grouping of values (displayed horizontally) of possibly different types, they are not named (nor is the compiler making any assumption of possible names).  Going further, tuples are different from arrays which are a grouping of values of the same type, not different types.

So, I believe there has to be some type of mechanisms in the language to map the tuple data structure to a different type of data structure.  The separate return values of a Vlang function (including other languages that use multiple returns) doesn't have such a mechanism.

So the below, should be expected to be illegal in Vlang:

```
struct Tuple
{
	data (int, int)
}
```

It should be expressed something like:

```
struct Tuple
{
	data1 int
	data2 int
}
```

A true tuple would show this as (int, int).  **The values are unnamed**.

A Vlang function with multiple separate returns, is just using the parentheses for visual grouping, not as defining a data structure.  

```
fn foo() (int, int) {
	return 2, 8
}
```

The Vlang way (also in some other languages) allows you to work with each returned value separately.  So we can just assign the 1st value to the variable and omit the 2nd value (by using the underscore "_").

`var, _ := foo() ` // var will only contain "2", and a separate variable needed for the other returned value.

/*
Attempting to assign multiple separate return values to a single variable is illegal in Vlang.
If the function was returning a tuple data structure, then legal, but that's a different language construct.
*/
`var := foo()` // illegal in Vlang, because in this case foo() returns multiple separate values (see above)

`(var1, var2)  := foo()` // also illegal in Vlang, since does not have tuples

`var1, var2 := foo()` // this is legal in Vlang

**Reply by timrichardson** [reply]
I am confused, this is true.
I think the current situation is not just confusing, but wrong. 

It is confusing because firstly this 

`mut tasks := []thread int{}`

is not very well documented. What is going on here? The array has elements of type thread, so what exactly is the annotation following? 
Without any reason otherwise, I deduced that the "annotation" type is a function return type. Certainly, you can not compile if your task function returns a type different to the "annoation" of `[]thread` (so in fact, the array has elements which are a handle to a function return type?)
So since it is legal to define a function that returns `(int,int)`  one would expect that `mut tasks := []thread (int, int) {}` is also legal. 

Explanations that "tuples" are not "first class types" is not very convincing if that is the final position of vlang. 

As to this:


fn foo() (int, int) {
	return 1,2
}

Formally, what is the return type? What else can you call it but a tuple? Whatever it is called, it is a type because functions have return types which matter in a strongly typed language. Note that decomposing the tuple by assignment (`a,b  = foo()`) is just syntax, I think so many languages do it now, it is rather conventional.

 If the compiler knows that this is wrong: `a,b,c = foo()` because foo() only returns (int,int) then I'd have to say that if it walks like a type and talks like a type, it's a type. 

So it seems there are two sets of types: one which is valid for a function, and a smaller set which is valid to annotate []thread 

This is why I think it is wrong. 

I am new to vlang, so if this is not a bug, something is not at all clear to me.

**Reply by Wajinn** [reply]
Tuple is a type of data structure, and the language has to support it.  The situation of Vlang, is like that of Go and a number of other languages.  It's kind of like saying, because I had this in my other language, it must be in this other language.  For example Vlang does concurrency a certain way or uses ":=", if a new user of C# demanded the language be changed to be like Vlang or Go, their people would balk at that.  Now, with Vlang being such a young language, there is much more leeway and this is seen as a feature request which can get consideration.

Maybe the Vlang developers can work support for tuples in, but clearly that would be their call and it can possibly be something with lots of unseen problems to even try to implement.  From the outside looking in, the best way I think they could go about it is to somehow map tuples as an anonymous struct with automatically generated names by the compiler.  Though I could be way off, as to the best way to implement such, if the developers even deemed this necessary to pursue.

> What else can you call it but a tuple?

But that's where we are getting stuck.  You can't call it a tuple, because that is a specific kind of data structure that contains various values which can be of different types, not a return of separate values.

The return type of below is **not** a tuple, despite how it might appear.  A tuple is not defined simply by appearance.
```
fn foo() (int, int) {
return 1,2
}
```
Below are presently illegal:

`(var1, var2) := foo()` // This is is illegal in Vlang, since tuple not supported.

`mut var := []thread (int, int) `// Thus this is also illegal, as is a tuple.

Below are presently legal:

`var1, var2 := foo()` // This is legal, because separate variables.

`mut var := []thread int{}` // This is legal, as single value.

`mut var := []thread []int{} `// This is also legal, as an array.


> ...if this is not a bug, something is not at all clear to me

It's not a bug, it's the design of this language, and others (like Go).  Maybe something can be figured out, in regards to tuples, but we should see it as a feature request (so is labeled correctly).

I do think a very good argument can be made that this should be clarified in the documentation, that Vlang functions with separate return values are very much not tuples.

**Comment by timrichardson** [comment]
From go documentation:
The (int, int) in this function signature explains that the return type is
two integers.

what can we say about vlang?

(I don't care if it is called a tuple or not. I did not use the term tuple, that was introduced by another commentor. To me it is a return type, that is clear, and it is written with parentheses. Sure it looks like a tuple, but we can call it a "multiple value return type" if you want. 

what I don't understand is that if it is a type, why can't I do this: []tuple (int, int) ?



On Mon, 1 Aug 2022 at 14:09, Wajinn ***@***.***> wrote:

> Tuple is a type of data structure, and the language has to support it. The
> situation of Vlang, is like that of Go and a number of other languages.
> It's kind of like saying, because I had this in my other language, it must
> be in this other language. For example Vlang does concurrency a certain way
> or uses ":=", if a new user of C# demanded the language be changed to be
> like Vlang or Go, their people would balk at that. Now, with Vlang being
> such a young language, there is much more leeway and this is seen as a
> feature request which can get consideration.
>
> Maybe the Vlang developers can work support for tuples in, but clearly
> that would be their call and it can possibly be something will lots of
> unseen problems to even try to implement. From the outside looking in, the
> best way I think they could go about it is to somehow map tuples as an
> anonymous struct with automatically generated names by the compiler. Though
> I could be way off, as to the best way to implement such, if the developers
> even deemed this necessary to pursue.
>
> What else can you call it but a tuple?
>
> But that's where we are getting stuck. You can't call it a tuple, because
> that is a specific kind of data structure that contains various values
> which can be of different types, not a return of separate values.
>
> The return type of below is *not* a tuple, despite how it might appear. A
> tuple is not defined simply by appearance.
>
> fn foo() (int, int) {
> return 1,2
> }
>
> Below are presently illegal:
>
> (var1, var2) := foo() // This is is illegal in Vlang, since tuple not
> supported.
>
> mut var := []thread (int, int) // Thus this is also illegal, as is a
> tuple.
>
> Below are presently legal:
>
> var1, var := foo() // This is legal, because separate variables.
>
> mut var := []thread int{} // This is legal, as single value.
>
> mut var := []thread []int{} // This is also legal, as an array.
>
> ...if this is not a bug, something is not at all clear to me
>
> It's not a bug, it's the design of this language, and others (like Go).
> Maybe something can be figured out, in regards to tuples, but we should see
> it as a feature request (so is labeled correctly).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/vlang/v/discussions/15300#discussioncomment-3293955>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHCNYRRHQRBJNCZOHNDJIDVW5EYTANCNFSM55F6BBGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 


*Tim Richardson CPA, Director, BA, BSc, MAcc*
GrowthPath. Finance transformation for SMEs via Cloud ERP, advanced
reporting, CRM

Mobile: +61 423 091 732 Office/Reception: +61 3 8678 1850.
Book call: https://vyte.in/growthpath/15
Timezone is Melbourne AU.
See this link for international time planning:
https://www.timeanddate.com/worldclock/meeting.html?year=2020&month=5&day=16&p1=152

GrowthPath Pty Ltd
ABN 18100392326
Xero Gold Partner. Dear Inventory, Zoho Analytics and Cin7 Implementation
Partner. Custom integration specialists.




<http://www.growthpath.com.au/>

**Reply by timrichardson** [reply]
Also, I am not a go coder, but I mucked around. 
This code runs. You can see that where a type is expected, I am able to use (int,int)

```
package main


import (
	"fmt"
)

func main() {

	var foo func(a int, b int) (int, int) = func(a int, b int) (int, int) {
		return a, b
	}

	var array_of_foo [1]func(a int, b int) (int, int)
	array_of_foo[0] = foo

	x, y := foo(5, 6)
	fmt.Println(x, y)
	w, v := array_of_foo[0](1, 2)
	fmt.Println(w, v)

}
```

Output:

```
5 6
1 2

Program exited.
```

**Comment by Wajinn** [comment]
> ...Also, I am not a go coder, but I mucked around...

I believe the Go code you have shown can be translated into Vlang, as follows:

```
type Foo = fn (a int, b int) (int, int) 

fn main() {
	mut array_of_foo := []Foo{}
	foo := fn (a int, b int) (int, int) {return a, b}
	
	array_of_foo << foo

	x, y := foo(5,6)
	println('$x, $y')
	
	w, v := array_of_foo[0](1, 2)
	println('$w, $v')
}

/*
results:
5, 6
1, 2
*/
```

**Comment by timrichardson** [comment]
Note that in go, the array has the full function signature.

In your example, you only say that array_of_foo hods a function variable.
Is that a function that returns nothing, or a function that returns
anything?
Either the type is wrong, or the type checking is wrong.



On Mon, 1 Aug 2022 at 22:55, Wajinn ***@***.***> wrote:

> ...Also, I am not a go coder, but I mucked around...
>
> I believe the Go code you have shown can be translated into Vlang, as
> follows:
>
> type Foo = fn (a int, b int) (int, int)
>
> fn main() {
> 	mut array_of_foo := []Foo{}
> 	foo := fn (a int, b int) (int, int) {return a, b}
> 	
> 	array_of_foo << foo
>
> 	x, y := foo(5,6)
> 	println('$x, $y')
> 	
> 	w, v := array_of_foo[0](1, 2)
> 	println('$w, $v')
> }
>
> /*
> results:
> 5, 6
> 1, 2
> */
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/vlang/v/discussions/15300#discussioncomment-3296910>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHCNYWNT7HHDU3ULIHYFGLVW7CMDANCNFSM55F6BBGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 


*Tim Richardson CPA, Director, BA, BSc, MAcc*
GrowthPath. Finance transformation for SMEs via Cloud ERP, advanced
reporting, CRM

Mobile: +61 423 091 732 Office/Reception: +61 3 8678 1850.
Book call: https://vyte.in/growthpath/15
Timezone is Melbourne AU.
See this link for international time planning:
https://www.timeanddate.com/worldclock/meeting.html?year=2020&month=5&day=16&p1=152

GrowthPath Pty Ltd
ABN 18100392326
Xero Gold Partner. Dear Inventory, Zoho Analytics and Cin7 Implementation
Partner. Custom integration specialists.




<http://www.growthpath.com.au/>

**Comment by timrichardson** [comment]
Sorry, I didn't realise that you defined a type.
Yes, I know that works.

My issue is that I can't write

[]thread (int,int)

but in go, I could. You said that vlang is like go in this respect, but it
is not. go's syntax is consistent.
I think we are going around in circles. You are not conceding my point. I
agree there is a workaround. But the language is not consistent.
There is a similar issue which has been marked as a bug so I hope it will
be fixed.


On Tue, 2 Aug 2022 at 06:55, Tim Richardson ***@***.***> wrote:

>
>
> Note that in go, the array has the full function signature.
>
> In your example, you only say that array_of_foo hods a function variable.
> Is that a function that returns nothing, or a function that returns
> anything?
> Either the type is wrong, or the type checking is wrong.
>
>
>
> On Mon, 1 Aug 2022 at 22:55, Wajinn ***@***.***> wrote:
>
>> ...Also, I am not a go coder, but I mucked around...
>>
>> I believe the Go code you have shown can be translated into Vlang, as
>> follows:
>>
>> type Foo = fn (a int, b int) (int, int)
>>
>> fn main() {
>> 	mut array_of_foo := []Foo{}
>> 	foo := fn (a int, b int) (int, int) {return a, b}
>> 	
>> 	array_of_foo << foo
>>
>> 	x, y := foo(5,6)
>> 	println('$x, $y')
>> 	
>> 	w, v := array_of_foo[0](1, 2)
>> 	println('$w, $v')
>> }
>>
>> /*
>> results:
>> 5, 6
>> 1, 2
>> */
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/vlang/v/discussions/15300#discussioncomment-3296910>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAHCNYWNT7HHDU3ULIHYFGLVW7CMDANCNFSM55F6BBGQ>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>
>
> --
>
>
> *Tim Richardson CPA, Director, BA, BSc, MAcc*
> GrowthPath. Finance transformation for SMEs via Cloud ERP, advanced
> reporting, CRM
>
> Mobile: +61 423 091 732 Office/Reception: +61 3 8678 1850.
> Book call: https://vyte.in/growthpath/15
> Timezone is Melbourne AU.
> See this link for international time planning:
> https://www.timeanddate.com/worldclock/meeting.html?year=2020&month=5&day=16&p1=152
>
> GrowthPath Pty Ltd
> ABN 18100392326
> Xero Gold Partner. Dear Inventory, Zoho Analytics and Cin7 Implementation
> Partner. Custom integration specialists.
>
>
>
>
> <http://www.growthpath.com.au/>
>


-- 


*Tim Richardson CPA, Director, BA, BSc, MAcc*
GrowthPath. Finance transformation for SMEs via Cloud ERP, advanced
reporting, CRM

Mobile: +61 423 091 732 Office/Reception: +61 3 8678 1850.
Book call: https://vyte.in/growthpath/15
Timezone is Melbourne AU.
See this link for international time planning:
https://www.timeanddate.com/worldclock/meeting.html?year=2020&month=5&day=16&p1=152

GrowthPath Pty Ltd
ABN 18100392326
Xero Gold Partner. Dear Inventory, Zoho Analytics and Cin7 Implementation
Partner. Custom integration specialists.




<http://www.growthpath.com.au/>

**Reply by Wajinn** [reply]
Maybe it would have been good to show an exact example of using []thread (int,int) in Go, thus it would have more clearly won the point.  But don't get me wrong, I'm not an adversary to the issue, just seeking clarity about it.  I do think the issue as a feature request is very valid.  

I'm sure the tuple issue, in some form or another, will keep popping up.  It also appears that it had troubled the Go community, and they created various tuple packages.  Such packages could probably be translated over or serve as inspiration.  So even if it wasn't directly built into the language, it could possibly be a module to add some kind of bi-directional tuple translation into a struct (if that even makes sense) or whatever solution eventually happens.

As it presently is, there are various workarounds (depending on what's being done) in terms of using arrays, multidimensional arrays, or structs (as shown by spytheman above).

**Reply by timrichardson** [reply]
It is not a tuple issue, it is a type issue. (int,int) is already a type in vlang, that ship has sailed. So it should be treated liked any other type.

**Reply by JalonSolov** [reply]
The main problem with tuples is that you lose information about the values.

With structs, the fields are named, and it doesn't matter what order you set the names in, and they're always accessible by those names.  It also doesn't matter if new fields are added, the fields are rearranged, etc.  As long as they keep the same names and types, it's all good.

With tuples, it is up to the caller to "know" which things are coming back, and in which order... and that all goes out the window if the routine that's returning the tuples changes the order, or inserts a new one in the middle, or...

**Reply by Apocryphon-X** [reply]
> [timrichardson](https://github.com/timrichardson): ...it is a type issue. (int,int) is already a type in vlang...

Looks like at this moment the compiler detects the 'multiple return value' as an `incomplete type` (using #15083 as reference), so maybe this thread could be considered a bug report _and_ a feature request at the same time.

**Reply by Apocryphon-X** [reply]
> it had troubled the Go community, and they created various tuple packages

Maybe this could fulfill the feature request part, but it would be cool to have some built-in support for tuples :-)
I was thinking about having named tuples like in python, but that will not work since V is compiled so nvm.

**Reply by Wajinn** [reply]
> The main problem with tuples is that you lose information about the values.
> 
> With structs, the fields are named, and it doesn't matter what order you set the names in, and they're always accessible by those names. It also doesn't matter if new fields are added, the fields are rearranged, etc. As long as they keep the same names and types, it's all good.
> 
> With tuples, it is up to the caller to "know" which things are coming back, and in which order... and that all goes out the window if the routine that's returning the tuples changes the order, or inserts a new one in the middle, or...

In doing a bit of research on this, I saw various suggestions and proposals on conversions from tuple to struct or return values to struct, where names were automatically generated in the struct and indexed so that they would be aligned to the left to right order of the return values.

**Comment by timrichardson** [comment]
#15083 is marked as a bug, so this may simply be a duplicate

**Reply by Apocryphon-X** [reply]
- New related issue: #15366

**Reply by Apocryphon-X** [reply]
- #15083 was closed leaving this as default behavior:
```failcompile
vlib/v/checker/tests/struct_multi_return_field_err.vv:2:16: error: cannot use multi return as field type
    1 | struct Tuple{
    2 |     data (int, int)
      |                   ^
    3 | }
    4 |
```
